### PR TITLE
fix(pillselectordropdown): support paste from context menu

### PR DESCRIPTION
**Pull request by booc0mtaco**
Due to how certain versions of IE11 handle paste and input events, sometimes the UI for `PillSelectorDropdown` instances would handle the pill creation **before** parsing the input. this would cause the field to be cleared before pills were created and would also destroy the user input. 

This change uses the [`Clipboard API`](https://caniuse.com/#search=clipboarddata) synthetic event to get the input data on paste, and refactors the pill parsing methods to use a parameter, so that it always uses the proper context and input value based on the current user workflow.

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @booc0mtaco has signed the CLA. Thanks for the pull request!

File: src/components/pill-selector-dropdown/PillSelectorDropdown.js [hunk 1]
```diff
@@ -177,21 +176,24 @@ class PillSelectorDropdown extends React.Component<Props, State> {
...
 
-    handlePaste = () => {
+    handlePaste = (event: SyntheticClipboardEvent<HTMLInputElement>) => {
         /**
          * NOTE (ishay): setTimeout is necessary because
          * otherwise addPillsFromInput gets triggered as soon
          * as the user "paste's", but before the inputValue
          * is actually updated.
          */
-        setTimeout(this.addPillsFromInput, 0);
```
Comment (alexkrolick) [inline src/components/pill-selector-dropdown/PillSelectorDropdown.js]: 👀 this is the problem

File: src/components/pill-selector-dropdown/__tests__/PillSelectorDropdown-test.js [hunk 2]
```diff
@@ -435,19 +441,68 @@ describe('components/pill-selector-dropdown/PillSelectorDropdown', () => {
...
-                .mock(instance)
-                .expects('addPillsFromInput')
-                .once();
+            instance.handlePaste(mockEvent);
+
+            expect(mockOnInput).toHaveBeenCalledWith(mockPastedValue, mockEvent);
+            expect(mockOnPillCreate).not.toHaveBeenCalled();
+        });
+
+        test('should call onPillCreate prop method with valid input', () => {
```
Comment (alexkrolick) [inline src/components/pill-selector-dropdown/__tests__/PillSelectorDropdown-test.js]: 👍

File: src/features/unified-share-modal/UnifiedShareModal.js [hunk 3]
```diff
@@ -363,6 +363,11 @@ class UnifiedShareModal extends React.Component<Props, State> {
                 }
             });
         }
+
+        this.setState({
```
Comment (jstoffan) [inline src/features/unified-share-modal/UnifiedShareModal.js]: This seems unrelated?

File: src/components/pill-selector-dropdown/PillSelectorDropdown.js [hunk 4]
```diff
@@ -177,21 +176,20 @@ class PillSelectorDropdown extends React.Component<Props, State> {
...
-    handlePaste = () => {
-        /**
-         * NOTE (ishay): setTimeout is necessary because
-         * otherwise addPillsFromInput gets triggered as soon
-         * as the user "paste's", but before the inputValue
-         * is actually updated.
-         */
-        setTimeout(this.addPillsFromInput, 0);
+    handlePaste = (event: SyntheticClipboardEvent<HTMLInputElement>) => {
+        event.nativeEvent.preventDefault();
```
Comment (jstoffan) [inline src/components/pill-selector-dropdown/PillSelectorDropdown.js]: Do we need to access `nativeEvent` here or does `event.preventDefault()` work?

File: src/components/pill-selector-dropdown/PillSelectorDropdown.js [hunk 5]
```diff
@@ -177,21 +176,20 @@ class PillSelectorDropdown extends React.Component<Props, State> {
...
+
+        const inputValue: string = event.clipboardData.getData('Text');
+        this.setState({ inputValue });
+        this.props.onInput(inputValue, event);
```
Comment (jstoffan) [inline src/components/pill-selector-dropdown/PillSelectorDropdown.js]: Is this going to cause issues with consuming expecting a `SyntheticInputEvent` rather than a `SyntheticClipboardEvent`? The flow type currently doesn't specify the parameters.

File: src/components/pill-selector-dropdown/PillSelectorDropdown.js [hunk 6]
```diff
@@ -177,21 +176,20 @@ class PillSelectorDropdown extends React.Component<Props, State> {
...
-         * NOTE (ishay): setTimeout is necessary because
-         * otherwise addPillsFromInput gets triggered as soon
-         * as the user "paste's", but before the inputValue
-         * is actually updated.
-         */
-        setTimeout(this.addPillsFromInput, 0);
+    handlePaste = (event: SyntheticClipboardEvent<HTMLInputElement>) => {
+        event.nativeEvent.preventDefault();
+
+        const inputValue: string = event.clipboardData.getData('Text');
```
Comment (jstoffan) [inline src/components/pill-selector-dropdown/PillSelectorDropdown.js]: Does this work in all browsers? Should it be `clipboardData.getData('text')` or `clipboardData.getData('text/plain')`, per [MDN](https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/getData)?

File: src/features/unified-share-modal/UnifiedShareModal.js (hunk 3 above)
Comment (booc0mtaco) [reply src/features/unified-share-modal/UnifiedShareModal.js]: true, will remove (came up in testing)

File: src/components/pill-selector-dropdown/PillSelectorDropdown.js (hunk 4 above)
Comment (booc0mtaco) [reply src/components/pill-selector-dropdown/PillSelectorDropdown.js]: It appears that the synthetic event works just as well, so i can switch to that

File: src/components/pill-selector-dropdown/PillSelectorDropdown.js (hunk 5 above)
Comment (booc0mtaco) [reply src/components/pill-selector-dropdown/PillSelectorDropdown.js]: The types around this are unfortunate :( BUT, the type for `handleInput` specifies an event or some object with `target`, which `SyntheticClipboarEvent` – and by extension its native cousin – has. So the minimum is checked here.

File: src/components/pill-selector-dropdown/PillSelectorDropdown.js (hunk 6 above)
Comment (booc0mtaco) [reply src/components/pill-selector-dropdown/PillSelectorDropdown.js]: React appears to normalize the parts of the event specific to the Clipboard API. There used to be more details about the `.types` documented, but they took those off. I'll double-check if a non-legacy MIME type works here (e.g., `text/plain`)

Comment (booc0mtaco) [reply src/components/pill-selector-dropdown/PillSelectorDropdown.js]: https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/types

Comment (booc0mtaco) [reply src/components/pill-selector-dropdown/PillSelectorDropdown.js]: `Text` and `text` work in all browsers, but IE11 doesn't support the full MIME types. Will lower-case this
